File: client-js/lib/runtime.js
/** Transport-agnostic request/response plumbing and SSE parsing.
 *
 * A line-by-line port of contree_client/runtime.py adapted to the
 * fetch platform: gzip decoding and connection pooling are handled by
 * fetch itself, so only the protocol logic lives here.
 */

export const CHUNK_SIZE = 65536;

export const PACKAGE_VERSION = "0.1.0";
export const UA_PRODUCT = `contree-client-js/${PACKAGE_VERSION}`;

const NODE_VERSION =
  typeof process !== "undefined" && process.versions?.node
    ? process.versions.node
    : null;
// browsers refuse to set the User-Agent header (a forbidden header
// name): the composed string is only ATTACHED as a header in Node,
// though userAgent() reports it everywhere
export const IS_NODE = NODE_VERSION !== null;
export const UA_RUNTIME =
  NODE_VERSION !== null ? `Node.js/${NODE_VERSION}` : "browser";
export const UA_PLATFORM =
  NODE_VERSION !== null
    ? `${process.platform}-${process.arch}`
    : (globalThis.navigator?.userAgent ?? "");

export const RETRY_DELAYS = Object.freeze([0.1, 0.2, 0.5, 1.0, 2.0, 5.0]);

// floor for reconnect loops that made no forward progress, so a server
// returning immediate empty streams does not spin the client
export const TIGHT_LOOP_FLOOR = 0.5;

/** An endless ladder of backoff delays: the ladder is walked once and
 * then the tail delay repeats forever. */
export function* retryDelays(delays = RETRY_DELAYS) {
  yield* delays;
  for (;;) {
    yield delays[delays.length - 1];
  }
}

/** Opt-in retries for transient failures of buffered requests.
 *
 * 425 (Too Early) and 429 (Too Many Requests) are a backend contract:
 * both mean the request was rejected before any processing, so
 * replaying them is always safe - even for a POST the caller hasn't
 * opted into unsafe retries for (see `call()`). */
export class RetryPolicy {
  constructor({
    statuses = [410, 425, 429],
    serverErrors = true,
    delays = RETRY_DELAYS,
    maxAttempts = 10,
    retryUnsafe = false,
  } = {}) {
    if (!delays.length) {
      throw new RangeError("RetryPolicy delays must not be empty");
    }
    for (const delay of delays) {
      if (!Number.isFinite(delay) || delay < 0) {
        throw new RangeError(
          `RetryPolicy delays must be finite and non-negative, got ${delay}`,
        );
      }
    }
    if (
      maxAttempts !== null &&
      (!Number.isInteger(maxAttempts) || maxAttempts < 1)
    ) {
      throw new RangeError("RetryPolicy maxAttempts must be an integer >= 1");
    }
    // defensive copies, frozen: a policy shared between clients must
    // not be mutable from the outside after validation
    this.statuses = Object.freeze([...statuses]);
    this.serverErrors = serverErrors;
    this.delays = Object.freeze([...delays]);
    this.maxAttempts = maxAttempts;
    this.retryUnsafe = retryUnsafe;
    Object.freeze(this);
  }

  retryableStatus(status) {
    if (this.statuses.includes(status)) {
      return true;
    }
    return this.serverErrors && status >= 500 && status < 600;
  }
}

/** Parse a Retry-After header: delta-seconds or an HTTP-date.
 * Negative values clamp to 0; anything unparsable reports null. */
export function parseRetryAfter(value) {
  if (value === null || value === undefined) {
    return null;
  }
  const seconds = Number(value);
  // Number("Infinity") parses: an infinite Retry-After would sleep
  // forever (or collapse into an immediate retry storm)
  if (Number.isFinite(seconds) && value.trim() !== "") {
    return Math.max(0, seconds);
  }
  const moment = Date.parse(value);
  if (Number.isNaN(moment)) {
    return null;
  }
  return Math.max(0, (moment - Date.now()) / 1000);
}

export function sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

export function monotonic() {
  return performance.now() / 1000;
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/** True if *ref* looks like an image/operation UUID. */
export function isUuid(ref) {
  return UUID_RE.test(ref);
}

export function quotePath(value) {
  return encodeURIComponent(String(value));
}

export function encodeQuery(query) {
  return Object.entries(query)
    .flatMap(([key, value]) =>
      (Array.isArray(value) ? value : [value]).map(
        (item) =>
          `${encodeURIComponent(key)}=${encodeURIComponent(item).replaceAll("%2F", "/")}`,
      ),
    )
    .join("&");
}

export function formatTimeParam(value) {
  if (value instanceof Date) {
    return value.toISOString();
  }
  return String(value);
}

/** Parse an ISO 8601 timestamp, tolerating nanosecond precision. */
export function parseDatetime(value) {
  // Date.parse only accepts millisecond precision: trim the tail
  const trimmed = value.replace(/(\.\d{3})\d+/, "$1");
  const moment = new Date(trimmed);
  if (Number.isNaN(moment.getTime())) {
    throw new RangeError(`unparsable datetime: ${value}`);
  }
  return moment;
}

const textDecoder = new TextDecoder();
const textEncoder = new TextEncoder();

export function bytesToText(bytes) {
  return textDecoder.decode(bytes);
}

export function textToBytes(text) {
  return textEncoder.encode(text);
}

export function bytesToBase64(bytes) {
  if (typeof Buffer !== "undefined") {
    return Buffer.from(bytes).toString("base64");
  }
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary);
}

export function base64ToBytes(value) {
  if (typeof Buffer !== "undefined") {
    return new Uint8Array(Buffer.from(value, "base64"));
  }
  const binary = atob(value);
  const bytes = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }
  return bytes;
}

/** The sha256 hexdigest of an upload payload.
 *
 * Accepts Uint8Array, string or Blob; a ReadableStream cannot be
 * re-read for the subsequent upload, so it reports null (the caller
 * skips deduplication). Blobs hash chunk by chunk in Node (10 GiB
 * must not materialize); browsers fall back to one arrayBuffer().
 */
export async function sha256(content) {
  if (typeof content === "string") {
    content = textToBytes(content);
  }
  if (content instanceof Uint8Array) {
    // the webcrypto global only appeared in Node 19: fall back to
    // node:crypto on Node 18 (browsers always have globalThis.crypto)
    const subtle = globalThis.crypto?.subtle;
    if (subtle === undefined) {
      const { createHash } = await import("node:crypto");
      return createHash("sha256").update(content).digest("hex");
    }
    const digest = await subtle.digest("SHA-256", content);
    return hex(new Uint8Array(digest));
  }
  if (typeof Blob !== "undefined" && content instanceof Blob) {
    if (NODE_VERSION !== null) {
      // node-only module, loaded lazily so browser bundlers never see it
      const { createHash } = await import("node:crypto");
      const digest = createHash("sha256");
      for await (const chunk of content.stream()) {
        digest.update(chunk);
      }
      return digest.digest("hex");
    }
    const digest = await globalThis.crypto.subtle.digest(
      "SHA-256",
      await content.arrayBuffer(),
    );
    return hex(new Uint8Array(digest));
  }
  return null;
}

function hex(bytes) {
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join(
    "",
  );
}

export function jsonBody(response) {
  return JSON.parse(bytesToText(response.body));
}

export function jsonObject(response) {
  const data = jsonBody(response);
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    const type =
      data === null ? "null" : Array.isArray(data) ? "array" : typeof data;
    throw new TypeError(`expected a JSON object, got ${type}`);
  }
  return data;
}

export function jsonArray(response) {
  const data = jsonBody(response);
  if (!Array.isArray(data)) {
    const type = data === null ? "null" : typeof data;
    throw new TypeError(`expected a JSON array, got ${type}`);
  }
  return data;
}

/** Extract diagnostic fields from an unsuccessful response. */
export function responseErrorDetails(response) {
  let error = bytesToText(response.body);
  let traceback = null;
  let payload = null;
  try {
    payload = JSON.parse(error);
  } catch {
    payload = null;
  }
  if (
    payload !== null &&
    typeof payload === "object" &&
    !Array.isArray(payload)
  ) {
    error = payload.error ?? error;
    if (Array.isArray(payload.traceback)) {
      traceback = payload.traceback.map(String);
    }
  }
  const parsedRetry = parseRetryAfter(response.headers["retry-after"] ?? null);
  const retryAfter = parsedRetry === null ? null : Math.trunc(parsedRetry);
  return { error, traceback, retryAfter };
}

/** Incremental sans-io parser for `text/event-stream` bytes.
 *
 * Feed raw chunks, get complete frames `{id, event, data}` back.
 * Comment lines (`: keepalive`) are discarded per the SSE spec.
 */
export class SSEParser {
  // a single SSE line/frame has no business being this large; the
  // cap keeps a misbehaving peer from growing the buffer unbounded
  static MAX_BUFFER = 4 * 1024 * 1024;

  constructor() {
    this.decoder = new TextDecoder();
    this.buffer = "";
    this.event = null;
    this.eventId = null;
    this.dataLines = [];
    this.pendingSize = 0;
    this.dirty = false;
  }

  feed(chunk) {
    this.buffer += this.decoder.decode(chunk, { stream: true });
    // the cap covers BOTH the unterminated line and the frame
    // accumulated so far: many short data lines must not grow the
    // pending event unboundedly
    if (this.buffer.length + this.pendingSize > SSEParser.MAX_BUFFER) {
      throw new RangeError(
        `SSE frame exceeds ${SSEParser.MAX_BUFFER} bytes before completion`,
      );
    }
    const frames = [];
    for (;;) {
      const newline = this.buffer.indexOf("\n");
      if (newline < 0) {
        break;
      }
      let line = this.buffer.slice(0, newline);
      this.buffer = this.buffer.slice(newline + 1);
      if (line.endsWith("\r")) {
        line = line.slice(0, -1);
      }
      if (!line) {
        const frame = this.flush();
        if (frame !== null) {
          frames.push(frame);
        }
        continue;
      }
      if (line.startsWith(":")) {
        continue;
      }
      const colon = line.indexOf(":");
      const name = colon < 0 ? line : line.slice(0, colon);
      let value = colon < 0 ? "" : line.slice(colon + 1);
      if (value.startsWith(" ")) {
        value = value.slice(1);
      }
      this.dirty = true;
      if (name === "id") {
        const parsed = Number.parseInt(value, 10);
        this.eventId = Number.isNaN(parsed) ? null : parsed;
      } else if (name === "event") {
        this.event = value;
      } else if (name === "data") {
        this.dataLines.push(value);
        this.pendingSize += value.length;
      }
    }
    return frames;
  }

  /** Return the pending frame, if any, and reset the state. */
  flush() {
    if (!this.dirty) {
      return null;
    }
    const frame = {
      id: this.eventId,
      event: this.event,
      data: this.dataLines.join("\n"),
    };
    this.event = null;
    this.eventId = null;
    this.dataLines = [];
    this.pendingSize = 0;
    this.dirty = false;
    return frame;
  }
}

/** Decode an SSE frame's JSON payload. */
export function decodeFramePayload(frame, lastEventId = null) {
  if (frame.event === "sse_error") {
    throw Object.assign(new Error(frame.data), { lastEventId });
  }
  if (!frame.data) {
    return null;
  }
  const payload = JSON.parse(frame.data);
  if (
    payload === null ||
    typeof payload !== "object" ||
    Array.isArray(payload)
  ) {
    return null;
  }
  return payload;
}

File: client/contree_client/runtime.py
"""Transport-agnostic request/response plumbing and SSE parsing."""

from __future__ import annotations

import asyncio
import hashlib
import json
import logging
import math
import re
import zlib
from collections.abc import AsyncIterator, Iterable, Iterator, Mapping, Sequence
from dataclasses import dataclass, field
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from functools import lru_cache
from importlib.metadata import PackageNotFoundError, distribution
from time import monotonic
from typing import IO, Any, cast
from urllib.parse import quote, urlencode

from .exceptions import (
    ERROR_CLASSES,
    APIStatusError,
    ServerError,
)
from .models import OperationEvent

CHUNK_SIZE = 65536

# The package logger: explicitly ERROR, so the library stays silent
# unless the user opts in (see contree_client.types.set_log_level).
logger = logging.getLogger("contree_client")
logger.setLevel(logging.ERROR)

SENSITIVE_HEADERS = frozenset(
    {
        "authorization",
        "proxy-authorization",
        "cookie",
        "set-cookie",
        "x-api-key",
        "x-auth-token",
    }
)

# Same policy the API applies to spawned process env (see the
# EventDataSpawn schema): any key ending with one of these suffixes is
# considered a secret, case-insensitively.
SENSITIVE_KEY_SUFFIXES = (
    "token",
    "secret",
    "password",
    "api_key",
    "credentials",
    "authorization",
)

REQUEST_DEADLINE_MESSAGE = "request deadline exceeded"


def remaining_timeout(
    deadline: float | None,
    maximum: float | None,
) -> float | None:
    """Return the smaller of the remaining deadline and *maximum*.

    Raise ``TimeoutError`` before transport I/O when the absolute
    monotonic deadline has elapsed.
    """
    if deadline is None:
        return maximum
    remaining = deadline - monotonic()
    if remaining <= 0:
        raise TimeoutError(REQUEST_DEADLINE_MESSAGE)
    if maximum is None:
        return remaining
    return min(remaining, maximum)


def redact_json(value: Any) -> Any:
    """Recursively replace values of secret-suffixed keys."""
    if isinstance(value, dict):
        return {
            key: "<redacted>"
            if str(key).lower().endswith(SENSITIVE_KEY_SUFFIXES)
            else redact_json(item)
            for key, item in value.items()
        }
    if isinstance(value, list):
        return [redact_json(item) for item in value]
    return value


class HeaderFormatter:
    """Lazy `%s` argument for logging HTTP headers, redacts secrets."""

    def __init__(self, headers: Iterable[tuple[str, str]]) -> None:
        self.headers = headers

    def __str__(self) -> str:
        redacted = {
            key: "<redacted>" if key.lower() in SENSITIVE_HEADERS else value
            for key, value in self.headers
        }
        return repr(redacted)


class BodyFormatter:
    """Lazy `%s` argument for logging HTTP bodies.

    JSON payloads are parsed and re-rendered with the values of
    secret-suffixed keys structurally redacted; other textual payloads
    are rendered as-is. Everything is truncated at *max_size*, binary
    payloads become a size marker and file-like bodies are never
    consumed.
    """

    def __init__(
        self,
        body: bytes | IO[bytes] | None,
        content_type: str = "",
        max_size: int = 4096,
    ) -> None:
        self.body = body
        self.content_type = content_type
        self.max_size = max_size

    def truncated(self, text: str, total: int) -> str:
        if len(text) > self.max_size:
            return f"{text[: self.max_size]}... <truncated, {total}B total>"
        return text

    def __str__(self) -> str:
        if self.body is None:
            return "<none>"
        if not isinstance(self.body, (bytes, bytearray)):
            return "<stream>"
        data = bytes(self.body)
        if not data:
            return "<empty>"
        # empty content type is treated as JSON: redaction must be the
        # default, not something a missing header can switch off
        if not self.content_type or "json" in self.content_type:
            try:
                payload = json.loads(data)
            except (ValueError, UnicodeDecodeError):
                # potentially credential-bearing: never echo raw bytes
                return f"<unparsable body {len(data)}B>"
            text = json.dumps(redact_json(payload), ensure_ascii=False)
            return self.truncated(text, len(data))
        if "text" in self.content_type:
            try:
                # decode the whole (already buffered) body so the
                # truncation marker actually fires for long payloads
                text = data.decode("utf-8")
            except UnicodeDecodeError:
                return f"<binary {len(data)}B>"
            return self.truncated(text, len(data))
        return f"<binary {len(data)}B Content-Type={self.content_type!r}>"


@dataclass
class RequestSpec:
    """A backend-agnostic description of a single API request."""

    method: str
    path: str
    query: Mapping[str, str | Sequence[str]] = field(default_factory=dict)
    headers: dict[str, str] = field(default_factory=dict)
    body: bytes | IO[bytes] | None = None
    content_type: str | None = None
    accept: str | None = None
    # retry safety: only idempotent requests may be replayed after a
    # lost response (a re-sent POST could spawn a second sandbox)
    idempotent: bool = False
    # Monotonic deadline passed to transport waits and retries. Sync buffered
    # reads can report expiry after the response completes.
    # None uses only the client's configured timeout.
    deadline: float | None = None
    # SSE only: bound the idle gap between reads. None keeps the
    # stream open indefinitely (the server sends keepalives); a
    # deadline-driven follower sets its remaining budget here so a
    # silent stream cannot outlive the caller's timeout
    read_timeout: float | None = None


@dataclass
class ResponseData:
    """A fully buffered response; header names are lower-cased."""

    status: int
    headers: dict[str, str]
    body: bytes


def distribution_version(name: str) -> str | None:
    """The installed distribution version via package metadata.

    Editable installs report ``"editable"`` (their recorded version
    is whatever was current at install time, i.e. a lie); missing
    metadata reports None.
    """
    try:
        dist = distribution(name)
    except PackageNotFoundError:
        return None
    raw = dist.read_text("direct_url.json")
    if raw:
        try:
            if json.loads(raw).get("dir_info", {}).get("editable"):
                return "editable"
        except ValueError:
            pass
    return dist.version


def package_version() -> str:
    return distribution_version("contree-client") or "unknown"


@lru_cache
def library_version(module: Any) -> str:
    """A ``name/version`` User-Agent token for a transport module.

    Version resolution: package metadata (``editable`` for editable
    installs), then the module's ``__version__``, then ``unknown`` -
    building the token must never fail the import.
    """
    name = module.__name__
    module_version = (
        distribution_version(name) or getattr(module, "__version__", None) or "unknown"
    )
    return f"{name}/{module_version}"


UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.ASCII | re.IGNORECASE,
)


def is_uuid(ref: str) -> bool:
    """True if *ref* looks like an image/operation UUID."""
    return UUID_RE.match(ref) is not None


def file_sha256(content: bytes | IO[bytes]) -> str | None:
    """The sha256 hexdigest of an upload payload.

    File-like payloads are hashed from their *current* position - the
    exact bytes a subsequent send would transmit - and the position is
    restored afterwards. A non-seekable stream cannot be re-read, so
    it reports None (the caller skips deduplication).
    """
    if isinstance(content, (bytes, bytearray)):
        return hashlib.sha256(content).hexdigest()
    if not content.seekable():
        return None
    start = content.tell()
    digest = hashlib.sha256()
    while chunk := content.read(CHUNK_SIZE):
        digest.update(chunk)
    content.seek(start)
    return digest.hexdigest()


RETRY_DELAYS = (0.1, 0.2, 0.5, 1.0, 2.0, 5.0)

# floor for reconnect loops that made no forward progress, so a server
# returning immediate empty streams does not spin the client
TIGHT_LOOP_FLOOR = 0.5


def retry_generator(delays: tuple[float, ...] = RETRY_DELAYS) -> Iterator[float]:
    """An endless ladder of backoff delays.

    Every ``next()`` returns a valid sleep time: the ladder is walked
    once and then the tail delay repeats forever, so callers never
    guard against ``StopIteration`` and never do index arithmetic. A
    finite retry budget is bounded externally (an attempt counter).
    """

    def generator() -> Iterator[float]:
        yield from delays
        while True:
            yield delays[-1]

    return generator()


@dataclass(frozen=True)
class RetryPolicy:
    """Opt-in retries for buffered request failures.

    Retries ``APIConnectionError`` and the configured HTTP statuses.
    ``server_errors`` extends this to every 5xx status. SSE consumers
    handle stream reconnection separately.
    """

    # 425 (Too Early) and 429 (Too Many Requests) are a backend
    # contract: both mean the request was rejected before any
    # processing, so replaying them is always safe - even for a POST
    # the caller hasn't opted into unsafe retries for (see `call()`)
    statuses: tuple[int, ...] = (410, 425, 429)
    server_errors: bool = True
    delays: tuple[float, ...] = RETRY_DELAYS
    # finite by default: unbounded retries are an explicit choice
    max_attempts: int | None = 10
    # non-idempotent requests (POST) are replayed only when the caller
    # explicitly accepts the double-execution risk - except 425/429,
    # which are always safe to replay (see `call()`)
    retry_unsafe: bool = False

    def __post_init__(self) -> None:
        if not self.delays:
            raise ValueError("RetryPolicy.delays must not be empty")
        for delay in self.delays:
            if not math.isfinite(delay) or delay < 0:
                raise ValueError(
                    f"RetryPolicy delays must be finite and non-negative, got {delay!r}"
                )
        if self.max_attempts is not None and (
            not isinstance(self.max_attempts, int)
            or isinstance(self.max_attempts, bool)
            or self.max_attempts < 1
        ):
            raise ValueError("RetryPolicy.max_attempts must be an integer >= 1")

    def retryable_status(self, status: int) -> bool:
        if status in self.statuses:
            return True
        return self.server_errors and 500 <= status < 600


def body_start(spec: RequestSpec) -> int | None:
    """The position a file-like body starts sending from.

    Captured before the first attempt so a retry replays exactly the
    same bytes; None for absent/bytes/non-seekable bodies.
    """
    body = spec.body
    if body is None or isinstance(body, (bytes, bytearray)):
        return None
    stream = cast("IO[bytes]", body)
    if not stream.seekable():
        return None
    return stream.tell()


def rewind_body(spec: RequestSpec, start: int | None) -> None:
    """Rewind a file-like request body to *start* before a retry."""
    body = spec.body
    if body is None or isinstance(body, (bytes, bytearray)):
        return
    stream = cast("IO[bytes]", body)
    if start is None or not stream.seekable():
        raise ValueError("cannot retry: streaming body is not seekable")
    stream.seek(start)


def parse_retry_after(value: str | None) -> float | None:
    """Parse a Retry-After header: delta-seconds or an HTTP-date.

    Negative values clamp to 0.0; anything unparsable reports None.
    """
    if value is None:
        return None
    try:
        seconds = float(value)
    except ValueError:
        seconds = None
    if seconds is not None:
        # float() happily parses "inf"/"nan": an infinite Retry-After
        # would make the retry loop sleep forever
        return max(0.0, seconds) if math.isfinite(seconds) else None
    try:
        moment = parsedate_to_datetime(value)
    except (TypeError, ValueError):
        return None
    if moment.tzinfo is None:
        moment = moment.replace(tzinfo=timezone.utc)
    return max(0.0, (moment - datetime.now(timezone.utc)).total_seconds())


def retry_after_delay(response: ResponseData) -> float | None:
    """The Retry-After header as seconds, when present and valid."""
    return parse_retry_after(response.headers.get("retry-after"))


def quote_path(value: str | int) -> str:
    return quote(str(value), safe="")


def encode_query(query: Mapping[str, str | Sequence[str]]) -> str:
    return urlencode(query, doseq=True, safe="/", quote_via=quote)


def format_time_param(value: str | int | float | datetime) -> str:
    if isinstance(value, datetime):
        return value.isoformat()
    return str(value)


def json_body(response: ResponseData) -> Any:
    return json.loads(response.body.decode("utf-8"))


def json_object(response: ResponseData) -> dict[str, Any]:
    data = json_body(response)
    if not isinstance(data, dict):
        raise TypeError(f"expected a JSON object, got {type(data).__name__}")
    return data


def json_array(response: ResponseData) -> list[Any]:
    data = json_body(response)
    if not isinstance(data, list):
        raise TypeError(f"expected a JSON array, got {type(data).__name__}")
    return data


def error_for_response(
    status_code: int,
    headers: Mapping[str, str],
    body: bytes,
) -> APIStatusError:
    """Build the exception matching an error response."""
    error: Any = body.decode("utf-8", "replace")
    traceback: list[str] | None = None
    try:
        payload = json.loads(body.decode("utf-8"))
    except (ValueError, UnicodeDecodeError):
        payload = None
    if isinstance(payload, dict):
        error = payload.get("error", error)
        raw_traceback = payload.get("traceback")
        if isinstance(raw_traceback, list):
            traceback = [str(item) for item in raw_traceback]
    parsed_retry = parse_retry_after(headers.get("retry-after"))
    retry_after = None if parsed_retry is None else int(parsed_retry)
    cls = ERROR_CLASSES.get(status_code)
    if cls is None:
        cls = ServerError if status_code >= 500 else APIStatusError
    return cls(
        status_code,
        error,
        traceback=traceback,
        retry_after=retry_after,
    )


class GzipStreamDecoder:
    """Incremental gzip decoder for streamed response bodies.

    Used by transports without built-in decompression (http.client);
    handles Z_SYNC_FLUSH-style streams, so decoded data comes out as
    soon as the server flushes it.
    """

    def __init__(self) -> None:
        self.state = zlib.decompressobj(16 + zlib.MAX_WBITS)

    def decompress(self, data: bytes) -> bytes:
        return self.state.decompress(data)

    def flush(self) -> bytes:
        tail = self.state.flush()
        # physical EOF must coincide with the gzip trailer: a stream
        # cut mid-flight would otherwise pass as a complete payload
        if not self.state.eof:
            raise EOFError(
                "compressed stream ended before the gzip trailer"
                " - the payload is truncated"
            )
        return tail


class IdentityStreamDecoder:
    """No-op decoder for uncompressed response bodies."""

    def decompress(self, data: bytes) -> bytes:
        return data

    def flush(self) -> bytes:
        return b""


def stream_decoder(
    content_encoding: str | None,
) -> GzipStreamDecoder | IdentityStreamDecoder:
    if content_encoding and content_encoding.strip().lower() == "gzip":
        return GzipStreamDecoder()
    return IdentityStreamDecoder()


def request_content(
    body: bytes | IO[bytes] | None,
) -> bytes | Iterator[bytes] | None:
    """Normalize a request body to something httpx accepts as content."""
    if body is None or isinstance(body, bytes):
        return body
    return iter(lambda: body.read(CHUNK_SIZE), b"")


def async_request_content(
    body: bytes | IO[bytes] | None,
) -> bytes | AsyncIterator[bytes] | None:
    """Normalize a request body for an asynchronous transport.

    A file-like body becomes an async iterator whose reads run in a
    worker thread, so blocking file I/O never stalls the event loop
    (and an async client never receives a sync iterator).
    """
    if body is None or isinstance(body, bytes):
        return body

    async def reader() -> AsyncIterator[bytes]:
        while chunk := await asyncio.to_thread(body.read, CHUNK_SIZE):
            yield chunk

    return reader()


@dataclass
class SSEFrame:
    """One parsed Server-Sent Events frame."""

    id: int | None = None
    event: str | None = None
    data: str = ""


class SSEParser:
    """Incremental sans-io parser for `text/event-stream` bytes.

    Feed raw chunks, get complete frames back. Comment lines
    (`: keepalive`) are discarded per the SSE specification.
    """

    # a single SSE line/frame has no business being this large; the
    # cap keeps a misbehaving peer from growing the buffer unbounded
    MAX_BUFFER = 4 * 1024 * 1024

    def __init__(self) -> None:
        self.buffer = bytearray()
        self.event: str | None = None
        self.event_id: int | None = None
        self.data_lines: list[str] = []
        self.pending_size = 0
        self.dirty = False

    def feed(self, chunk: bytes) -> list[SSEFrame]:
        self.buffer += chunk
        # the cap covers BOTH the unterminated line and the frame
        # accumulated so far: many short data lines must not grow the
        # pending event unboundedly
        if len(self.buffer) + self.pending_size > self.MAX_BUFFER:
            raise ValueError(
                f"SSE frame exceeds {self.MAX_BUFFER} bytes before completion"
            )
        frames: list[SSEFrame] = []
        while True:
            newline = self.buffer.find(b"\n")
            if newline < 0:
                break
            raw = bytes(self.buffer[:newline])
            del self.buffer[: newline + 1]
            line = raw.decode("utf-8", "replace").removesuffix("\r")
            if not line:
                frame = self.flush()
                if frame is not None:
                    frames.append(frame)
                continue
            if line.startswith(":"):
                continue
            name, _, value = line.partition(":")
            value = value.removeprefix(" ")
            self.dirty = True
            if name == "id":
                try:
                    self.event_id = int(value)
                except ValueError:
                    self.event_id = None
            elif name == "event":
                self.event = value
            elif name == "data":
                self.data_lines.append(value)
                self.pending_size += len(value)
        return frames

    def flush(self) -> SSEFrame | None:
        """Return the pending frame, if any, and reset the state."""
        if not self.dirty:
            return None
        frame = SSEFrame(
            id=self.event_id,
            event=self.event,
            data="\n".join(self.data_lines),
        )
        self.event = None
        self.event_id = None
        self.data_lines = []
        self.pending_size = 0
        self.dirty = False
        return frame


def decode_event_frame(
    frame: SSEFrame,
    last_event_id: int | None = None,
) -> OperationEvent | None:
    """Decode an SSE frame into an OperationEvent.

    Return None for frames that carry no event payload.
    """
    if frame.event == "sse_error":
        error = ConnectionError(frame.data)
        error.__dict__["last_event_id"] = last_event_id
        raise error
    if not frame.data:
        return None
    payload = json.loads(frame.data)
    if not isinstance(payload, dict):
        return None
    return OperationEvent.from_dict(payload)
